java-spiffe-core/src/test/java/io/spiffe/spiffeid/SpiffeIdTest.java

```diff
@@ -1,89 +1,136 @@
 
-        assertAll("spiffeId",
-                () -> assertEquals("trust-domain.org", spiffeId.getTrustDomain().toString()),
-                () -> assertEquals("/path1/path2", spiffeId.getPath())
-        );
+        assertEquals("spiffe://trust-domain.org/path1/path2/path3", spiffeIdToString);
     }
 
-    @Test
-    void of_TrustDomainAndNoPaths_ReturnsSpiffeIdWithTrustDomain() {
-        val trustDomain = TrustDomain.of("trust-domain.org");
-
-        val spiffeId = SpiffeId.of(trustDomain);
+    @ParameterizedTest
+    @MethodSource("provideTestValidSpiffeIds")
+    void testParseValidSpiffeId(String input, TrustDomain expectedTrustDomain, String expectedPath) {
+        SpiffeId result;
+        try {
+            result = SpiffeId.parse(input);
+            assertEquals(expectedTrustDomain, result.getTrustDomain());
+            assertEquals(expectedPath, result.getPath());
+        } catch (Exception e) {
+            fail("Unexpected error", e);
+        }
+    }
 
-        assertAll("spiffeId",
-                () -> assertEquals("trust-domain.org", spiffeId.getTrustDomain().toString()),
-                () -> assertEquals("", spiffeId.getPath())
+    static Stream<Arguments> provideTestValidSpiffeIds() {
+        return Stream.of(
+                Arguments.of("spiffe://example.org", TrustDomain.of("example.org"), ""),
+                Arguments.of("spiffe://example.org/path", TrustDomain.of("example.org"), "/path"),
+                Arguments.of("spiffe://example.org/path1/path2", TrustDomain.of("example.org"), "/path1/path2"),
+                Arguments.of("spiffe://example.org/PATH1/PATH2", TrustDomain.of("example.org"), "/PATH1/PATH2"),
+                Arguments.of("SPIFFE://EXAMPLE.ORG/path1/path2", TrustDomain.of("example.org"), "/path1/path2"),
+                Arguments.of("spiffe://example.org/p!a$t&h'/(e)l*e+m,e;n=t", TrustDomain.of("example.org"), "/p!a$t&h'/(e)l*e+m,e;n=t"),
+                Arguments.of("spiffe://example.org/p@th", TrustDomain.of("example.org"), "/p@th"),
+                Arguments.of("spiffe://example.org/pa:th/element:", TrustDomain.of("example.org"), "/pa:th/element:"),
+                Arguments.of("  spiffe://domain.test/path1/path2  ", TrustDomain.of("domain.test"), "/path1/path2"),
+                Arguments.of("spiffe://example.org/9eebccd2-12bf-40a6-b262-65fe0487d453", TrustDomain.of("example.org"), "/9eebccd2-12bf-40a6-b262-65fe0487d453")
         );
     }
 
-    @Test
-    void of_TrustDomainAndPathsWithCaps_ReturnsSpiffeIdNormalized() {
-        val trustDomain = TrustDomain.of("TRuST-DoMAIN.Org");
-
-        val spiffeId = SpiffeId.of(trustDomain, "PATH1", "paTH2");
+    @ParameterizedTest
+    @MethodSource("provideInvalidSpiffeIds")
+    void testParseInvalidSpiffeId(String input, String expected) {
+        SpiffeId result;
+        try {
+            result = SpiffeId.parse(input);
+            assertEquals(expected, result.toString());
```

Comment:
should this be a `fail()`? are there cases where parsing invalid spiffe ids don't result in an exception?

Reply:
good catch!